src/rawposix/src/fs_calls.rs

```diff
@@ -987,3 +989,616 @@ pub fn fcntl_syscall(
+    if kernel_fd == -1 {
+        return syscall_error(Errno::EFAULT, "sync_file_range", "Invalid Cage ID");
+    } else if kernel_fd == -9 {
+        return syscall_error(Errno::EBADF, "sync", "Bad File Descriptor");
+    }
+
+    let ret = unsafe {
+        libc::sync_file_range(kernel_fd, offset, nbytes, flags)
+    };
+
+    if ret < 0 {
+        let errno = get_errno();
+        return handle_errno(errno, "sync_file_range");
+    }
+    ret
+}
+
+//------------------------------------READLINKAT SYSCALL------------------------------------
+/*
+ *  `readlinkat` reads the value of a symbolic link.
+ *  Reference: https://man7.org/linux/man-pages/man2/readlinkat.2.html
+ *
+ *  ## Arguments:
+ *   - `virtual_fd`: File descriptor for directory (or AT_FDCWD for current working directory).
+ *   - `path`: Path to the symbolic link to read.
+ *   - `buf`: Buffer to store the link target.
+ *   - `buflen`: Size of the buffer.
+ *
+ *  ## Implementation Details:
+ *   - Handles both AT_FDCWD and explicit directory file descriptors.
+ *   - Converts virtual file descriptor to kernel file descriptor using `convert_fd_to_host`.
+ *   - Converts paths using `convpath` and `normpath` for proper path handling.
+ *   - Adds LIND_ROOT prefix to create full kernel path.
+ *   - Removes LIND_ROOT prefix from result before copying to user buffer.
+ *
+ *  ## Return Value:
+ *   - Number of bytes placed in `buf` on success.
+ *   - `-1` on failure, with `errno` set appropriately.
+ */
+pub fn readlinkat_syscall(
+    cageid: u64,
+    virtual_fd_arg: u64,
+    virtual_fd_cageid: u64,
+    path_arg: u64,
+    path_cageid: u64,
+    buf_arg: u64,
+    buf_cageid: u64,
+    buflen_arg: u64,
+    buflen_cageid: u64,
+    arg8: u64,
+    arg9: u64,
+    arg10: u64,
+    arg11: u64,
+) -> i32 {
+    let virtual_fd = sc_convert_sysarg_to_i32(virtual_fd_arg, virtual_fd_cageid, cageid);
+    let path = sc_convert_path_to_host(path_arg, path_cageid, cageid);
+    let buf = sc_convert_uaddr_to_host(buf_arg, buf_cageid, cageid) as *mut u8;
+    let buflen = sc_convert_sysarg_to_usize(buflen_arg, buflen_cageid, cageid);
+
+    let mut libcret;
+    let mut path = path.to_string_lossy().to_string();
+    let libc_buflen = buflen + LIND_ROOT.len();
+    let mut libc_buf = vec![0u8; libc_buflen];
+    if virtual_fd == libc::AT_FDCWD {
+        // Check if the fd is AT_FDCWD
+        let cage_arc = get_cage(cageid).unwrap();
+        let cwd_container = cage_arc.cwd.read();
+        path = format!("{}/{}", cwd_container.to_str().unwrap(), path);
+        // Convert the path from relative path (lind-wasm perspective) to real kernel path (host kernel
+        // perspective)
+        let relpath = normpath(convpath(&path), cageid);
+        let relative_path = relpath.to_str().unwrap();
+        let full_path = format!("{}{}", LIND_ROOT, relative_path);
+        let c_path = std::ffi::CString::new(full_path).unwrap();
```

Comment:
I dislike that we're doing this in trusted code because there really isn't a reason to do so.  I'd much prefer this lived in a grate.  However that grate currently doesn't exist.  So, if need be, we can fix this later by moving the code then.